File: resources/php/login2.php
<?php

session_start();

date_default_timezone_set("Asia/Kolkata");

require_once __DIR__ . '/mail/Exception.php';
require_once __DIR__ . '/mail/PHPMailer.php';
require_once __DIR__ . '/mail/SMTP.php';


require_once __DIR__ . '/config.php';
require_once __DIR__ . '/smtp.php';

try {
    // Create connection
    $conn = new mysqli($servername, $username, $password, $dbname);

    // Check connection
    if ($conn->connect_error) {
        throw new Exception("Connection failed: " . $conn->connect_error);
    }

    // Check if the form is submitted
    if ($_SERVER["REQUEST_METHOD"] == "POST") {

        if (isset($_POST["username"]) && isset($_POST["pwd"])) {
            // Retrieve form data
            $usernamept = htmlspecialchars(trim($_POST["username"]));
            $passwordpt = htmlspecialchars(trim($_POST["pwd"]));

            $checkQuery = "SELECT adminid, username, password, email FROM admin WHERE username = ?";
            if (!($checkStmt = $conn->prepare($checkQuery))) {
                throw new Exception("Prepare failed: (" . $conn->errno . ") " . $conn->error);
            }
            if (!$checkStmt->bind_param("s", $usernamept)) {
                throw new Exception("Binding parameters failed: (" . $checkStmt->errno . ") " . $checkStmt->error);
            }
            if (!$checkStmt->execute()) {
                throw new Exception("Execute failed: (" . $checkStmt->errno . ") " . $checkStmt->error);
            }
            $checkStmt->store_result();
            $checkStmt->bind_result($adminid, $username, $password, $email);
            $checkStmt->fetch();

            if ($checkStmt->num_rows() > 0) {
                if (($usernamept == $username) && ($password !== null) && (password_verify($passwordpt, $password))) {

                    // Store admin data in the session
                    $_SESSION['adminid'] = $adminid;
                    $_SESSION['username'] = $username;
                    $_SESSION['email'] = $email;

                    $otppt = str_pad(random_int(0, 999999), 6, '0', STR_PAD_LEFT);

                    $mail = new \PHPMailer\PHPMailer\PHPMailer(true);

                    try {
                        //Server settings
                        $mail->SMTPDebug = 0;
                        $mail->isSMTP();
                        $mail->Host = $smtphost;
                        $mail->SMTPAuth = true;
                        $mail->Username = $smtpusername;
                        $mail->Password = $smtppassword;
                        $mail->SMTPSecure = \PHPMailer\PHPMailer\PHPMailer::ENCRYPTION_STARTTLS;
                        $mail->Port = $smtpport;

                        //Recipients
                        $mail->setFrom($smtpusername, 'Foodelight');
                        $mail->addAddress($email);

                        //Content
                        $mail->isHTML(true);
                        $mail->Subject = 'Login Authentication for Admin';
                        $mail->Body    = "Hi " . $username . ",<br><br>Your OTP for Login is: " . $otppt . "<br><br>This OTP is valid for 2 minutes only.<br><br>Please use this OTP to login to Foodelight.<br><br>Thanks,<br>Foodelight";

                        $result = $mail->send();
                        if ($result == 1) {
                            echo 'OTP has been sent to your email';

                            // Insert OTP for this adminid in admin_otp table
                            $expiresAt = date('Y-m-d H:i:s', time() + 120);
                            $isUsed = 0;
                            $checkQuery1 = "INSERT INTO admin_otp(adminid, otp, expires_at, is_used) VALUES (?, ?, ?, ?)";
                            if (!($checkStmt1 = $conn->prepare($checkQuery1))) {
                                throw new Exception("Prepare failed: (" . $conn->errno . ") " . $conn->error);
                            }
                            if (!$checkStmt1->bind_param("iisi", $adminid, $otppt, $expiresAt, $isUsed)) {
                                throw new Exception("Binding parameters failed: (" . $checkStmt1->errno . ") " . $checkStmt1->error);
                            }
                            if (!$checkStmt1->execute()) {
                                throw new Exception("Error: " . $checkStmt1->error);
                            }
                            $checkStmt1->close();

                        } else {
                            echo "ERROR";
                        }

                    } catch (Exception $e) {
                        echo 'Message could not be sent. Mailer Error: ', $mail->ErrorInfo;
                    }
                } else {
                    echo "Invalid credentials";
                }

            } else {
                echo "Invalid credentials";
            }

        } else if (isset($_POST["otp"])) {
            try {
                $adminid = $_SESSION['adminid'] ?? null;
                if (!$adminid) {
                    throw new Exception("Admin ID missing from session.");
                }
                $checkStmt = $conn->prepare("SELECT id FROM admin_otp WHERE adminid = ? AND otp = ? AND is_used = 0 AND NOW() <= expires_at");
                if ($checkStmt === false) {
                    throw new Exception($conn->error);
                }
                $checkStmt->bind_param("ii", $adminid, $_POST["otp"]);
                $checkStmt->execute();
                $result = $checkStmt->get_result();

                if ($result->num_rows > 0) {
                    $row = $result->fetch_assoc();
                    $otpId = $row['id'];

                    $updateStmt = $conn->prepare("UPDATE admin_otp SET is_used = 1 WHERE id = ?");
                    if ($updateStmt === false) {
                        throw new Exception($conn->error);
                    }
                    $updateStmt->bind_param("i", $otpId);
                    $updateStmt->execute();

                    echo "Admin Login Success";

                } else {
                    echo "Invalid OTP!";
                }
            } catch (Exception $e) {
                die("Error: " . $e->getMessage());
            }
        } else {
            throw new Exception("Error: One or more form fields are missing.");
        }
    }

    // Close the database connection
    $conn->close();
} catch (Exception $e) {
    echo  $e->getMessage(), "\n";
}
?>

File: resources/php/login2.py
import random
import hmac
from datetime import datetime, timedelta, timezone

from flask import request, session
from werkzeug.security import check_password_hash, generate_password_hash

from db import get_connection
from db_init import initialize_schema
from mailer import send_html_mail


def register_routes(app):
    @app.route("/resources/php/login2.py", methods=["POST"])
    def login2_php():
        conn = None
        try:
            conn = get_connection()
            initialize_schema(conn)

            if request.method == "POST":
                form = request.form

                if "username" in form and "pwd" in form:
                    usernamept = (form.get("username", "") or "").strip()
                    passwordpt = (form.get("pwd", "") or "").strip()

                    if usernamept == "" or passwordpt == "":
                        return "Error: All fields are required."

                    with conn.cursor() as cur:
                        if usernamept.isdigit():
                            cur.execute(
                                """
                                SELECT adminid, username, password, email
                                FROM admin
                                WHERE LOWER(username) = LOWER(%s)
                                   OR LOWER(email) = LOWER(%s)
                                   OR adminid = %s
                                LIMIT 1
                                """,
                                (usernamept, usernamept, int(usernamept)),
                            )
                        else:
                            cur.execute(
                                """
                                SELECT adminid, username, password, email
                                FROM admin
                                WHERE LOWER(username) = LOWER(%s)
                                   OR LOWER(email) = LOWER(%s)
                                LIMIT 1
                                """,
                                (usernamept, usernamept),
                            )
                        row = cur.fetchone()

                    stored_password = row.get("password", "") if row else ""
                    password_ok = False
                    used_legacy_plain = False

                    if row and stored_password:
                        try:
                            password_ok = check_password_hash(stored_password, passwordpt)
                        except ValueError:
                            password_ok = False

                        if not password_ok and hmac.compare_digest(stored_password, passwordpt):
                            password_ok = True
                            used_legacy_plain = True

                    if row and password_ok:
                        adminid = row.get("adminid")
                        username = row.get("username", "")
                        email = row.get("email", "")

                        session["adminid"] = adminid
                        session["username"] = username
                        session["email"] = email

                        otppt = str(random.randint(0, 999999)).zfill(6)
                        body = (
                            "Hi "
                            + username
                            + ",<br><br>Your OTP for Login is: "
                            + otppt
                            + "<br><br>This OTP is valid for 2 minutes only.<br><br>Please use this OTP to login to Foodelight.<br><br>Thanks,<br>Foodelight"
                        )
                        sent, error = send_html_mail(email, "Login Authentication for Admin", body)
                        if sent:
                            expires_at = datetime.now(timezone.utc) + timedelta(minutes=2)
                            with conn.cursor() as cur:
                                cur.execute(
                                    "INSERT INTO admin_otp(adminid, otp, expires_at, is_used) VALUES (%s, %s, %s, %s)",
                                    (adminid, otppt, expires_at, 0),
                                )
                                # Upgrade legacy plain-text admin passwords after a successful login.
                                if used_legacy_plain:
                                    cur.execute(
                                        "UPDATE admin SET password = %s WHERE adminid = %s",
                                        (generate_password_hash(passwordpt), adminid),
                                    )
                            conn.commit()
                            return "OTP has been sent to your email"
                        return "ERROR" if not error else "Message could not be sent. Mailer Error: " + error

                    return "Invalid credentials"

                if "otp" in form:
                    try:
                        adminid = session.get("adminid")
                        if not adminid:
                            raise Exception("Admin ID missing from session.")
                        otp = (form.get("otp", "") or "").strip()

                        with conn.cursor() as cur:
                            cur.execute(
                                "SELECT id, expires_at FROM admin_otp WHERE adminid = %s AND otp = %s AND is_used = 0",
                                (adminid, otp),
                            )
                            row = cur.fetchone()

                        if row:
                            expires_at = row.get("expires_at")
                            if expires_at is None:
                                return "Invalid OTP!"
                            if expires_at.tzinfo is None:
                                expires_at = expires_at.replace(tzinfo=timezone.utc)
                            if datetime.now(timezone.utc) > expires_at:
                                return "OTP expired!"

                            with conn.cursor() as cur:
                                cur.execute("UPDATE admin_otp SET is_used = 1 WHERE id = %s", (row.get("id"),))
                            conn.commit()
                            return "Admin Login Success"

                        return "Invalid OTP!"
                    except Exception as exc:
                        return "Error: " + str(exc)

                raise Exception("Error: One or more form fields are missing.")

            return ""
        except Exception as exc:
            return str(exc) + "\n"
        finally:
            if conn is not None:
                conn.close()

